gen/internals.py

```diff
@@ -237,138 +372,234 @@ def all_conditions_met(setter):
+        if resolvable.is_error:
+            # TODO(cmaloney): Should re-raise the original error with it's original context.
+            raise SkipError(
+                "Value depended upon {} has an error: {}".format(resolvable.name, resolvable.error),
+                [(name, copy.copy(self._eval_stack))])
+
+        return resolvable.value
 
     def _calculate_target(self, target):
-        def evaluate_var(name):
-            try:
-                self._get(name)
-            except CalculatorError as ex:
-                log.debug("Error calculating %s: %s. Chain: %s", name, ex.message, ex.chain)
+        finalized_arguments = dict()
 
+        # TODO(cmaloney): All the arguments depended upon by the arguments resolved here should be
+        # included in the target's full set of finalized arguments.
         for name in target.variables:
-            evaluate_var(name)
+            self._ensure_finalized(self._arguments[name])
 
         for name, sub_scope in target.sub_scopes.items():
-            if name not in self._arguments:
-                evaluate_var(name)
+            self._ensure_finalized(self._arguments[name])
+            resolvable = self._arguments[name]
 
-            # If the internal arg is None, there was an error, don't check if it
-            # is a legal choice.
-            if self._arguments[name] is None:
+            assert resolvable.is_finalized, " _resolve_name should have resulted in finalization " \
+                "of {}".format(resolvable)
+
+            # Tried solving for the condition but couldn't, so we can't check
+            # the sub-scope because we don't know which one to check.
+            if resolvable.is_error:
                 continue
 
-            choice = self._get(name)
+            assert resolvable.is_resolved, "uhhh: {}".format(resolvable)
 
-            if choice not in sub_scope.cases:
-                self._errors[name] = "Invalid choice {}. Must choose one of {}".format(
-                    choice, ", ".join(sorted(sub_scope.keys())))
-                continue
+            # Must be in the cases since we add validation functions for all switches automatically.
+            assert resolvable.value in sub_scope.cases
 
-            self._calculate_target(sub_scope.cases[choice])
+            # Calculate all the items in the sub-scope
+            sub_target = sub_scope.cases[resolvable.value]
+            self._calculate_target(sub_target)
 
+            # This .update() is safe because Resolver guarantees each argument
+            # only ever has one value / resolvable.
+            finalized_arguments.update(sub_target.arguments)
+
+        target.finalize(finalized_arguments)
+
+    # Force calculation of all arguments by accessing the arguments in this
+    # scope and recursively all sub-scopes.
+    def resolve(self):
+        assert not self._resolved, "Resolvers should only be resolved once"
+        self._resolved = True
```

Comment:
So there is a function which looks just like that: https://github.com/dcos/dcos/blob/master/gen/internals.py#L374. The "Resolver" class is an internal implementation detail of it. No one other than that one function, resolve_configuration, should ever see / deal with an unresolved Resolver class.

as far as why a class: I think it makes a lot easier to track the common state, as well as environment (things like ArgumentStack, context, and all the validation functions). They all are located together, and easily accessible by the members.

The resolver class once resolved, then provides a very precise interface to retrieve how the resolution ran (the status_dict), as well as all arguments which were calculated (.arguments), and ideally even indexing into the targets it was given eventually (currently it's easier for callers to just look at the resolver objects they passed in because the python objects are passed by reference. Eventually I want to attach names to the targets, which should make for better error messages).

Another advantage of the class, is that when something goes wrong, all the state of the world is centralized, which makes it a lot easier to print out what that state is, rather than having multiple inter-dependent objects, or trying to track / pass it through as multiple function arguments (or even one big function argument passed to a half dozen functions). This also means that when we find configuration errors (things like the unset configuration, or when a validation function fails), that all the state / world is available and around

The class also takes multiple things in, and gives multiple things out. The goal is that the resolver has a simple API to always use it. It keeps track of partially complete state (If an error happens halfway through resolving, it could be inspected / probed because all that state is there). When there are issues in resolving configuration, the bits are always all there. This in particular I think will be helpful to resolve: https://github.com/dcos/dcos/blob/master/dcos_installer/async_server.py#L116. Doing that requires we keep track of what caused what, and can always trace back to the individual fields present in the configuration ui of the web installer.

Arguably, could make the resolve() call part of the constructor, which would give you more of the API you're asking about, but from my perspective it is an important different change to the class from just setting up the environment required for resolving.